### DISCUSSION TITLE
ERROR: Can not find path - thought I moved library to correct locations

### DISCUSSION THREAD

**Discussion by adw1011** [pyinstaller/pyinstaller#5872] [Help]
I thought fixing this error would be fairly simple, but apparently that is not the case.  Running Mac OS 11.4 on Apple M1, Python 3.9.4, PyInstaller 4.3.  I thought I put libshiboken6.abi3.6.0.dylib in the right places (/usr/local/lib and on DYLD_LIBRARY_PATH).  What am I missing?

17261 ERROR: Can not find path ./libshiboken6.abi3.6.0.dylib (needed by /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/PySide6/QtSql.abi3.so)
17264 ERROR: Can not find path ./libshiboken6.abi3.6.0.dylib (needed by /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/PySide6/QtNetwork.abi3.so)
17269 ERROR: Can not find path ./libshiboken6.abi3.6.0.dylib (needed by /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/PySide6/QtCore.abi3.so)
17273 ERROR: Can not find path ./libshiboken6.abi3.6.0.dylib (needed by /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/PySide6/QtGui.abi3.so)
17278 ERROR: Can not find path ./libshiboken6.abi3.6.0.dylib (needed by /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/PySide6/QtWidgets.abi3.so)
17346 ERROR: Can not find path ./libshiboken6.abi3.6.0.dylib (needed by /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/PySide6/libpyside6.abi3.6.0.dylib)

**Comment by rokm** [comment]
Missing shiboken errors should be mostly harmless - at least that's the case `PySide2` (you can make a `onedir` build, and check that libshiboken ends up collected into dist), and the program should work in spite of them.

However, PySide6 is not yet supported by PyInstaller.

**Reply by rokm** [reply]
If you're feeling adventurous, you can try out #5865 - it will still complain about missing `libshiboken6`, but should otherwise work.

**Reply by rokm** [reply]
Ah, wait, you're on M1 - that's not supported yet, either (unless you're running x86_64 version of python on Rosetta). 

So you'd need both #5581 and #5865.

**Reply by adw1011** [reply]
Thank you for the quick, accurate replies.  I should have seen the PySide6 support.  I don't know how I missed that.  From a post earlier this year I thought the M1 arch was supported.  On the M1, terminal windows can be opened using Rosetta.  Would you know if opening the terminal in Rosetta solves the architecture issue for PyInstaller?  If it does, then I can try out #5865.

rokm, not sure my skill level is up to doing what you suggested in terms of #5865 and #5581.   Am I correct in assuming that I would have to pull two branches of the dev tree and merge them?  

Pardon any mistakes in my parlance!

**Reply by rokm** [reply]
> Would you know if opening the terminal in Rosetta solves the architecture issue for PyInstaller? If it does, then I can try out #5865.

The best way to check is to run python and check the output of 
```python
import platform
platform.machine()
```
I think you'll need `macos11`/`universal2` version of python for `x86_64` mode to be available. You can also try running `arch -x86_64 python [...]` to force specific arch slice out of multi-arch fat binary.

But, even if you manage to get python running in `x86_64` mode, you'll also need to ensure that python packages you're using (e.g., `PySide6`) are of `x86_64` variant. It's probably best to create a new virtual environment (where you're careful about forcing x64_64 arch to all python/pip commands).

It might be easier to stick with `arm64`, though.

> rokm, not sure my skill level is up to doing what you suggested in terms of #5865 and #5581. Am I correct in assuming that I would have to pull two branches of the dev tree and merge them?

Yeah, you should create a new branch from current `develop`, then merge onto it the branch from #5865 and from #5581 (hopefully there are no conflicts...). Once you have that ready, you need to rebuild the bootloader, and then install the whole thing by pointing (arch-forced) `pip install` at the top-level directory of the checked out repository with rebuilt bootloader.

**Reply by adw1011** [reply]
I had forgotten I have a Mac Mini (Intel) with my app running with PySide6.  I think I tried #5865 correctly.  I brought down the code and built the bootloader (python ./waf all).  Then I went up a folder level and did a python setup.py install.  The version on the PyInstaller is 5.0.dev0.  I ran that on my Mac Mini and still get the same errors as I originally posted. Command line was 'PyInstaller --onefile --windowed myapp.py'. 

I also wanted to add that if you check the Rosetta box in 'Get Info' on the Terminal app, the platform.machine command gives you x86_64 instead of arm64.

**Reply by rokm** [reply]
> I ran that on my Mac Mini and still get the same errors as I originally posted.

Sure, you're expected to get the `Can not find path ./libshiboken6.abi3.6.0.dylib ` messages during the build (like I said, that doesn't change). The question is, does the frozen application work? If not, what's the run-time error message?

---

As a side note, why do we fail to find the shiboken library? Because PySide6 extensions (e.g., `site-packages/PySide6/QtCore.abi3.so`) link against it as `@rpath/libshiboken6.abi3.6.1.dylib`, and they set their rpath to `@loader_path` and `@loader_path/Qt/lib`. However, the `libshiboken6.abi3.6.1.dylib` is not actually in either of those locations - it is located in `site-packages/shiboken6/libshiboken6.abi3.6.1.dylib`.  

So why does it work in python? Because importing `PySide6` imports `shiboken6.Shiboken` early on, and `Shiboken` extension (`site-packages/shiboken6/Shiboken.abi3.so`) loads the `libshiboken6.abi3.6.1.dylib` shared library.

And that's why PyInstaller's dependency analysis cannot find the `libshiboken6.abi3.6.1.dylib` shared library when analyzing PySide6 extensions (which results in those error messages - which should probably be downgraded to warnings to avoid confusing users); however, we add a hidden import on `shiboken6`, end up collecting `shiboken6.Shiboken` extension, and when analyzing that, we pick up the `libshiboken6.abi3.6.1.dylib` shared library as its dependency (and this time, it is found because it is linked as `@rpath/libshiboken6.abi3.6.1.dylib` and rpath is `@loader_path` and the library is in same folder as the extension) . So in the end, it still works.

The same applies to PySide2.

**Reply by adw1011** [reply]
My app is still crashing.  I looked at the crash report and the console log.  I don't know if this is the relevant portion of the console log but this looked relevant(?):

default	18:51:14.507756-0400	syspolicyd	temporarySigning type=1 matchFlags=0x0 path=/var/folders/1c/qx3m2sss257f7jhpck8p_hdh0000gn/T/_MEICWsrAB/pandas/_libs/ops_dispatch.cpython-39-darwin.so
default	18:51:14.508183-0400	syspolicyd	MacOS error: -67062
error	18:51:14.531514-0400	tccd	FAIL: PID[70933]: SecTaskCopySigningIdentifier(): [22: Invalid argument]
error	18:51:14.531820-0400	tccd	FAIL: PID[70933]: SecTaskCopySigningIdentifier(): [22: Invalid argument]
default	18:51:14.531939-0400	tccd	PID[319] is checking access for target PID[70933]
error	18:51:14.532051-0400	tccd	FAIL: PID[70933]: SecTaskCopySigningIdentifier(): [22: Invalid argument]
default	18:51:14.534460-0400	tccd	MacOS error: -67062
error	18:51:14.536020-0400	tccd	Failed to copy signing info for 70933, responsible for file:///Users/xxx/python/dist/gwpc.app/Contents/MacOS/gwpc: #-67062: Error Domain=NSOSStatusErrorDomain Code=-67062 "(null)"
error	18:51:14.538731-0400	tccd	FAIL: PID[70933]: SecTaskCopySigningIdentifier(): [22: Invalid argument]
default	18:51:14.538922-0400	tccd	Service kTCCServiceDeveloperTool does not allow prompting; returning denied.
error	18:51:14.540127-0400	syspolicyd	Unable to get certificates array: (null)
default	18:51:14.540698-0400	syspolicyd	MacOS error: -67062
error	18:51:14.541987-0400	syspolicyd	Error Domain=NSOSStatusErrorDomain Code=-67062

**Reply by rokm** [reply]
Can you try running non-app-bundle version from terminal, and see if there's any run-time errors printed there?

**Reply by adw1011** [reply]
I ran my program from the command line and there are no runtime errors.  Last night I thought it might be related to my app not having permission to run because of the 'SecTaskCopySigningIdentifier' in the error listing above, but giving the app authorization (assuming that I properly authorized it) did not fix the problem.

**Reply by rokm** [reply]
From your console log, it does look like a signature (or lack thereof?) problem. Is your Mac set to allow running only apps from app store (or perhaps signed apps, and you did not sign your bundle)? At any rate, this does not seem related to PySide6 at all... but just to be sure, can you try building and running a `tkinter`-based hello world app?

**Reply by adw1011** [reply]
I ran a little tkinter based app and it does work (Mac Mini -x86 arch, Mac OS 10.15.7, Python 2.7.6 and 3.95).  Not sure what tk version.  Anyway, I ran it using '#!/usr/local/bin/python3' as the first line in the simple python program so it would pick up the latest python version.  That worked.  I also took out that line, so I assume it picked up the older Apple version on my machine which is 2.7.16.  That version also worked.  The one thing that seemed odd (but might be because I've never used pyinstaller before) is that after I double-click on the app icon, it launches and then seems to die but then relaunches and runs.

**Reply by rokm** [reply]
> The one thing that seemed odd (but might be because I've never used pyinstaller before) is that after I double-click on the app icon, it launches and then seems to die but then relaunches and runs.

Are you perhaps building your .app bundle in `onefile` mode? What is the (full) PyInstaller command you are using?

**Reply by adw1011** [reply]
pyinstaller --onefile --windowed test_tk.py

I thought --onefile was a default anyway.  Do you think my real app with Qt/PySide6 has to be signed somehow?

**Reply by rokm** [reply]
> pyinstaller --onefile --windowed test_tk.py
> 
> I thought --onefile was a default anyway.

I suspect `onefile` is causing the apparent die-and-relaunch behavior that you mentioned. `onedir` is the default build mode, and it's probably a better choice for an .app bundle. Especially if you plan to codesign it, as you'll have no way to sign the embedded binaries/libraries otherwise (unless you use `--codesign-identity` option with #5581).

> Do you think my real app with Qt/PySide6 has to be signed somehow?

Could be, but perhaps that depends on the settings of your Mac. As far as I'm aware (I'm not really all that familiar with macOS, though), x86_64 usually does not require signed binaries, but arm64 does. (There could be potentially an issue with `universal2` python 3.9.x libraries and non-#5581 PyInstaller and invalidation of signatures on collected libraries, but I don't think I've seen that before).

I suppose the easiest way to tell would be if you temporarily disable GateKeeper on your Mac, and see if that's gets your app running. 

On the side note, does your real application use pandas? Because your console log seems to indicate there was a signature-related issue on `pandas/_libs/ops_dispatch.cpython-39-darwin.so` file, with the temporary file path that would correspond to `onefile` build. If that's the case, perhaps you can try building a basic Hello World PySide6 application and see if that works?

**Reply by adw1011** [reply]
I fully disabled gatekeeper and tried again.  No joy, but I found thisQt plugin error.  Does this suggest anything to you or is it just a misleading error message?

gwpc	This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
default	14:42:02.058152-0400	kernel	AMFI: Process does not have a code signature, allowing core dump for pid 6917
default	14:42:02.058417-0400	runningboardd	[executable<gwpc(501)>:6917] Death sentinel fired!

**Reply by adw1011** [reply]
Ignore last msg.  Think its a dead end.

**Reply by adw1011** [reply]
I built a simple application with PySide6 and it still fails to launch using pyinstaller.  However, I ran the same application using PySide2 and it launches correctly.  The version of my pyinstaller says 5.0.dev0 (for both PySide2 and PySide6 tests).  These were run on my Mac Mini x86 arch.

**Reply by rokm** [reply]
Can you attach the build log for PySide6 version of the simple application?

**Reply by adw1011** [reply]
Here it is.
[py6_out.txt](https://github.com/pyinstaller/pyinstaller/files/6580306/py6_out.txt)

**Reply by rokm** [reply]
I don't see any mentions of the new `hook-PySide6*` hooks being ran. Are you sure you're using #5865 here?

**Reply by rokm** [reply]
This is how the relevant part of build log looks on my test machine:

<details><summary>Log excerpt</summary>
<p>

```
6348 INFO: Building Tree Tree-02.toc
6353 INFO: Loading module hook 'hook-encodings.py' from '/Users/rok/Development/pyi-pyside6/venv/lib/python3.9/site-packages/PyInstaller/hooks'...
6435 INFO: Loading module hook 'hook-distutils.util.py' from '/Users/rok/Development/pyi-pyside6/venv/lib/python3.9/site-packages/PyInstaller/hooks'...
6439 INFO: Loading module hook 'hook-pickle.py' from '/Users/rok/Development/pyi-pyside6/venv/lib/python3.9/site-packages/PyInstaller/hooks'...
6441 INFO: Loading module hook 'hook-heapq.py' from '/Users/rok/Development/pyi-pyside6/venv/lib/python3.9/site-packages/PyInstaller/hooks'...
6445 INFO: Loading module hook 'hook-difflib.py' from '/Users/rok/Development/pyi-pyside6/venv/lib/python3.9/site-packages/PyInstaller/hooks'...
6447 INFO: Loading module hook 'hook-PySide6.QtWidgets.py' from '/Users/rok/Development/pyi-pyside6/venv/lib/python3.9/site-packages/PyInstaller/hooks'...
6904 ERROR: Can not find path ./libshiboken6.abi3.6.1.dylib (needed by /Users/rok/Development/pyi-pyside6/venv/lib/python3.9/site-packages/PySide6/QtWidgets.abi3.so)
6993 INFO: Loading module hook 'hook-multiprocessing.util.py' from '/Users/rok/Development/pyi-pyside6/venv/lib/python3.9/site-packages/PyInstaller/hooks'...
6997 INFO: Loading module hook 'hook-PySide6.QtCore.py' from '/Users/rok/Development/pyi-pyside6/venv/lib/python3.9/site-packages/PyInstaller/hooks'...
7005 ERROR: Can not find path ./libshiboken6.abi3.6.1.dylib (needed by /Users/rok/Development/pyi-pyside6/venv/lib/python3.9/site-packages/PySide6/QtCore.abi3.so)
7014 INFO: Loading module hook 'hook-PySide6.py' from '/Users/rok/Development/pyi-pyside6/venv/lib/python3.9/site-packages/PyInstaller/hooks'...
7061 INFO: Loading module hook 'hook-sysconfig.py' from '/Users/rok/Development/pyi-pyside6/venv/lib/python3.9/site-packages/PyInstaller/hooks'...
7072 INFO: Loading module hook 'hook-xml.py' from '/Users/rok/Development/pyi-pyside6/venv/lib/python3.9/site-packages/PyInstaller/hooks'...
7114 INFO: Loading module hook 'hook-distutils.py' from '/Users/rok/Development/pyi-pyside6/venv/lib/python3.9/site-packages/PyInstaller/hooks'...
7115 INFO: Loading module hook 'hook-PySide6.QtGui.py' from '/Users/rok/Development/pyi-pyside6/venv/lib/python3.9/site-packages/PyInstaller/hooks'...
7119 ERROR: Can not find path ./libshiboken6.abi3.6.1.dylib (needed by /Users/rok/Development/pyi-pyside6/venv/lib/python3.9/site-packages/PySide6/QtGui.abi3.so)
7145 INFO: Loading module hook 'hook-PySide6.QtNetwork.py' from '/Users/rok/Development/pyi-pyside6/venv/lib/python3.9/site-packages/PyInstaller/hooks'...
7149 ERROR: Can not find path ./libshiboken6.abi3.6.1.dylib (needed by /Users/rok/Development/pyi-pyside6/venv/lib/python3.9/site-packages/PySide6/QtNetwork.abi3.so)
7226 INFO: Looking for ctypes DLLs
```

</p>
</details>

**Reply by adw1011** [reply]
I'm not sure I have the right version.  I think I brought down #5865 and then followed the instructions in the readme file.  I included the version number at the top of the file I attached.  Since I never worked with Github, maybe I didn't grab everything I needed.  If you're not seeing the hooks I assume I made a mistake.

**Reply by rokm** [reply]
Hmm, the version itself doesn't tell us anything, because `5.0.dev0` will be reported by `develop` branch and all branches stemming from it.

Try creating a new, clean virtual environment:
```
python3.9 -m venv test-venv
. test-venv/bin/activate
pip install https://github.com/rokm/pyinstaller/archive/refs/heads/qt6-wip.zip
pip install PySide6

# Check that the installed version contains new hooks
ls test-venv/lib/python3.9/site-packages/PyInstaller/hooks/hook-PySide6*

# Build the test program
pyinstaller [...]
```

**Reply by adw1011** [reply]
My PySide6 test program works on the Mac Mini with your pyinstaller.  I also moved over the executable to my  Apple M1 machine and it will execute there as well.  I assume that once the M1 machine has the executable, it uses Rosetta to launch it.  My app, which is fairly large and probably written poorly since I am new to Python will not execute on the Mini.  The way I import files or maybe some other issue is probably creating problems. 

rokm, thank you for all the help you have given me.  I'll have to devise some strategy to isolate what is causing the failure.  It dies instantly, so that's potentially a clue.  It looks like pyinstaller apps with your PySide6 fixes will run on M1 if they are copied over.  I think they can also be built on the M1 as is, but the architecture switch in Terminal would have to be set and the Python installation would need to be all x86.

**Reply by rokm** [reply]
I would suggest that you forget about app bundles (the `--windowed` argument) for the time being, and focus on getting the classic `onedir` build running on Mac Mini (or on M1 in x86_64 mode). App bundles are difficult to debug because they lack console/terminal, and you cannot see the exception information when your application terminates due to unhandled exception.

The regular onedir build needs to be run from terminal, but will allow you to see exception information. Once you have that completely working, try making the .app bundle with --windowed switch. (Or alternatively, implement top-level exception handling in your application that will log exceptions to a pre-determined file on disk. That way, you can check for exception and its traceback when your application "dies instantly"). It could be a missing module that did not get collected automatically, or a package missing a data file (these are also not collected automatically, unless we have a hook for the package) - but you need to see what the exception is, so you can narrow the problem down.

**Reply by adw1011** [reply]
Thanks for the suggestions.

**Reply by adw1011** [reply]
When do you think your PySide6 changes will become production?

**Reply by adw1011** [reply]
I did get the 'onefile' to work.  I type the name of my application at the command line and it launches and runs properly.  Trying to double-click on the windowed version still doesn't work, but I wanted to at least pass along that my larger program is running off the PySide6 pyinstaller.

**Reply by rokm** [reply]
> Trying to double-click on the windowed version still doesn't work, but I wanted to at least pass along that my larger program is running off the PySide6 pyinstaller.

If non-app-bundle works but app-bundle does not, try putting everything in your entry point script into `try/except` block, and write the exception information (or even better, whole traceback via `traceback` module) into a text file in a pre-determined location (e.g., user's home directory)

**Reply by adw1011** [reply]
I have the windowed app launching now but many of the windows don't function properly.  The 'onefile' version seems to work ok.  I did a lot of code cleanup. Using traceback I was able to find certain statements that I had to move out of __init__ blocks.
